### PR TITLE
Allow translation of HOL string type to CakeML string type

### PR DESCRIPTION
**Pull request by myreen**
The new feature is off by default, but importantly it is used 
for the strings in the AST and tokens that the lexer produces.

Closes issue #666.

### CODE REVIEW

**Comment by myreen** [comment]
I realised that there is a cheat in the type soundness proof. As a result, I'm closing this until that cheat is removed.

**Comment by myreen** [comment]
Commit b326ba64894cea56f2b521f1f448b2dacdda8312 removed the last cheat.

File: compiler/backend/proofs/flat_exh_matchProofScript.sml [hunk 1]
```diff
@@ -467,6 +467,7 @@ Theorem do_app_thm:
      state_rel ctors t1 t2 /\
      do_app cc s2 op vs2 = SOME (t2, r2)
 Proof
+
```
Comment (xrchz) [inline compiler/backend/proofs/flat_exh_matchProofScript.sml]:
```suggestion
```

File: compiler/backend/proofs/flat_to_patProofScript.sml [hunk 2]
```diff
@@ -466,12 +471,21 @@ Proof
...
+  \\ fs [patSemTheory.list_to_v_def,flatSemTheory.list_to_v_def]
+QED
+
 val do_app = Q.prove(
   `∀cc co op vs s0 s res.
      do_app b s0 op vs = SOME (s,res)
      ⇒
      do_app (compile_state co cc s0) (Op op) (compile_vs vs) =
        SOME (compile_state co cc s,map_result compile_v compile_v res)`,
+
```
Comment (xrchz) [inline compiler/backend/proofs/flat_to_patProofScript.sml]:
```suggestion
```

File: compiler/backend/proofs/flat_uncheck_ctorsProofScript.sml [hunk 3]
```diff
@@ -259,6 +259,7 @@ val do_app_correct = Q.prove (
      ∃r' s2'. do_app F s1' op vs' = SOME (s2', r') ∧
               s_rel s2 s2' ∧
               result_rel v_rel v_rel r r'`,
+
```
Comment (xrchz) [inline compiler/backend/proofs/flat_uncheck_ctorsProofScript.sml]:
```suggestion
```

**Comment by xrchz** [comment]
A good followup to this would be trying to remove the use of `mlstring` as much as possible in HOL definitions that are translated, if I understand correctly the new functionality.

**Pull request merged**